components/scream/src/physics/shoc/shoc_pblintd_height_impl.hpp

```diff
@@ -13,12 +13,123 @@ namespace shoc {
 
 template<typename S, typename D>
 KOKKOS_FUNCTION
-void Functions<S,D>::pblintd_height(const Int& shcol, const Int& nlev, const uview_1d<const Spack>& z, const uview_1d<const Spack>& u, const uview_1d<const Spack>& v, const uview_1d<const Spack>& ustar, const uview_1d<const Spack>& thv, const uview_1d<const Spack>& thv_ref, const uview_1d<Spack>& pblh, const uview_1d<Spack>& rino, const uview_1d<bool>& check)
+void Functions<S,D>::pblintd_height(const MemberType& team, const Int& nlev, const Int& npbl, 
+        const uview_1d<const Spack>& z, const uview_1d<const Spack>& u, 
+        const uview_1d<const Spack>& v, const Scalar& ustar, const uview_1d<const Spack>& thv, 
+        const Scalar& thv_ref, Scalar& pblh, const uview_1d<Spack>& rino, bool& check)
 {
-  // TODO
-  // Note, argument types may need tweaking. Generator is not always able to tell what needs to be packed
-}
+  //
+  // PBL height calculation:  Scan upward until the Richardson number between
+  // the first level and the current level exceeds the "critical" value.
+  //
+  const Scalar tiny = 1.e-36;
+  const Scalar fac  = 100.;
+  const Scalar ricr  =  0.3;
+  const Scalar ggr = C::gravit;
+
+  Spack z_plus, z_mid, rino_plus, rino_mid;
+  Spack vvk, pblh_sp, rino_sp;
+
+  Int index = 0;
```

Comment:
@jgfouca , yes, serial code is very simple, it find the first level exceeds the "critical" value, and then calculate the pblh via interpolation of rino (richardson number). and pblh is also ependent on latest updated rino. To parallel this piece of code requires parallel reduction calculation of richardson number of the level index, and then use this index for pblh and rino calculation.

The second and third parallel_for presumbly can be merged to 1 parallel_for, but there is sqeuence dependence of rino and pblh, so I separate them into 2 parallel_for loops.

Reply:
@xyuan , can you maybe add some comments with what you've described above, so that it's easier for others to understand how the parallelism affects this particular function?

Reply:
> @xyuan , can you maybe add some code comments with what you've described above, so that it's easier for others to understand how the parallelism affects this particular function?

yes, I added the comments to the code.

Reply:
Great, thanks!